pkg/api/db.go

```diff
@@ -0,0 +1,230 @@
+*/
+
+package api
+
+import (
+	"errors"
+	"fmt"
+
+	"time"
+
+	log "github.com/golang/glog"
+	c "github.com/opensds/opensds/pkg/context"
+	"github.com/opensds/opensds/pkg/db"
+	"github.com/opensds/opensds/pkg/model"
+	"github.com/opensds/opensds/pkg/utils"
+	"github.com/opensds/opensds/pkg/utils/constants"
+	"github.com/satori/go.uuid"
+)
+
+func CreateVolumeDBEntry(ctx *c.Context, in *model.VolumeSpec) (*model.VolumeSpec, error) {
+	if in.Id == "" {
+		in.Id = uuid.NewV4().String()
+	}
+	if in.Size <= 0 {
+		errMsg := fmt.Sprintf("Invalid volume size: %d", in.Size)
+		log.Error(errMsg)
+		return nil, errors.New(errMsg)
+	}
+	if in.AvailabilityZone == "" {
+		log.Warning("Use default availability zone when user doesn't specify availabilityZone.")
+		in.AvailabilityZone = "default"
+	}
+	if in.CreatedAt == "" {
+		in.CreatedAt = time.Now().Format(constants.TimeFormat)
+	}
+	vol := &model.VolumeSpec{
+		BaseModel: &model.BaseModel{
+			Id:        in.Id,
+			CreatedAt: in.CreatedAt,
+		},
+		UserId:           in.UserId,
+		TenantId:         in.TenantId,
+		Name:             in.Name,
+		Description:      in.Description,
+		Size:             in.Size,
+		AvailabilityZone: in.AvailabilityZone,
+		Status:           model.VOLUME_CREATING,
+	}
+	result, err := db.C.CreateVolume(ctx, vol)
+	if err != nil {
+		log.Error("When add volume to db:", err)
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func ExtendVolumeDBEntry(ctx *c.Context, volID string) (*model.VolumeSpec, error) {
+	volume, err := db.C.GetVolume(ctx, volID)
+	if err != nil {
+		log.Error("Get volume failed in extend volume method: ", err)
+		return nil, err
+	}
+
+	if volume.Status != model.VOLUME_AVAILABLE {
+		errMsg := "The status of the volume to be extended must be available"
+		log.Error(errMsg)
+		return nil, errors.New(errMsg)
+	}
+	volume.Status = model.VOLUME_EXTENDING
+	// Store the volume data into database.
+	result, err := db.C.ExtendVolume(ctx, volume)
+	if err != nil {
+		log.Error("When extend volume in db module:", err)
+		return nil, err
+	}
+	return result, nil
+}
+
+func CreateVolumeAttachmentDBEntry(ctx *c.Context, in *model.VolumeAttachmentSpec) (*model.VolumeAttachmentSpec, error) {
+	vol, err := db.C.GetVolume(ctx, in.VolumeId)
+	if err != nil {
+		log.Error("Get volume failed in create volume attachment method: ", err)
+		return nil, err
+	}
+	if vol.Status != model.VOLUME_AVAILABLE {
+		errMsg := "Only the status of volume is available, attachment can be created"
+		log.Error(errMsg)
+		return nil, errors.New(errMsg)
+	}
+	if in.Id == "" {
+		in.Id = uuid.NewV4().String()
+	}
+	if in.CreatedAt == "" {
+		in.CreatedAt = time.Now().Format(constants.TimeFormat)
+	}
+	if len(in.AdditionalProperties) == 0 {
+		in.AdditionalProperties = map[string]interface{}{"attachment": "attachment"}
+	}
+	if len(in.ConnectionData) == 0 {
+		in.ConnectionData = map[string]interface{}{"attachment": "attachment"}
+	}
+	if in.Platform == "" {
```

Comment:
Why do you add these assertion? From my point of view, it has no meaning to assert these empty fields.

Reply:
There is no need to implement async process of Update method. It updates directly at the API level without going though the backend.